src/daft-local-execution/src/dispatcher.rs

```diff
@@ -1,74 +1,175 @@
+        morsel_size: Option<usize>,
     ) -> DaftResult<()> {
         let mut next_worker_idx = 0;
-        let mut send_to_next_worker = |data: PipelineResultType| {
+        let mut send_to_next_worker = |data: Arc<MicroPartition>| {
             let next_worker_sender = worker_senders.get(next_worker_idx).unwrap();
             next_worker_idx = (next_worker_idx + 1) % worker_senders.len();
             next_worker_sender.send(data)
         };
 
-        let mut buffer = RowBasedBuffer::new(self.morsel_size);
-        while let Some(morsel) = receiver.recv().await {
-            if morsel.should_broadcast() {
-                for worker_sender in &worker_senders {
-                    if worker_sender.send(morsel.clone()).await.is_err() {
+        for receiver in input_receivers {
+            let mut buffer = morsel_size.map(RowBasedBuffer::new);
+            while let Some(morsel) = receiver.recv().await {
+                if let Some(buffer) = &mut buffer {
+                    buffer.push(&morsel);
+                    if let Some(ready) = buffer.pop_enough()? {
+                        for r in ready {
+                            if send_to_next_worker(r).await.is_err() {
+                                return Ok(());
+                            }
+                        }
+                    }
+                } else if send_to_next_worker(morsel).await.is_err() {
+                    return Ok(());
+                }
+            }
+            // Clear all remaining morsels
+            if let Some(buffer) = &mut buffer {
+                if let Some(last_morsel) = buffer.pop_all()? {
+                    if send_to_next_worker(last_morsel).await.is_err() {
                         return Ok(());
                     }
                 }
-            } else {
-                buffer.push(morsel.as_data());
-                if let Some(ready) = buffer.pop_enough()? {
-                    for r in ready {
-                        if send_to_next_worker(r.into()).await.is_err() {
-                            return Ok(());
+            }
+        }
+        Ok(())
+    }
+}
+
+impl Dispatcher for RoundRobinDispatcher {
+    fn dispatch(
+        &self,
+        input_receivers: Vec<CountingReceiver>,
+        num_workers: usize,
+        runtime_handle: &mut ExecutionRuntimeHandle,
+        name: &'static str,
+    ) -> Vec<Receiver<Arc<MicroPartition>>> {
+        let (worker_senders, worker_receivers): (Vec<_>, Vec<_>) =
+            (0..num_workers).map(|_| create_channel(1)).unzip();
+        let morsel_size = self.morsel_size;
+        runtime_handle.spawn(
+            Self::dispatch_inner(worker_senders, input_receivers, morsel_size),
+            name,
+        );
+        worker_receivers
+    }
+}
+
+/// A dispatcher that distributes morsels to workers in an unordered fashion.
+/// Used if the operator does not require maintaining the order of the input.
+pub(crate) struct UnorderedDispatcher {
```

Comment:
Script that tests for uneven runtimes across tasks. With the `UnorderedDispatcher`, performance is considerably increased.
```
daft.context.set_execution_config(default_morsel_size=1)

@daft.udf(return_dtype=daft.DataType.int64())
def do_work(x):
    x = x.to_pylist()
    # my laptop has 12 cores, so 12 workers will be spawned to run this udf.
    if x[0] % 12 == 0:
        print("doing a lot of work on ", x)
        time.sleep(1)
    else:
        print("doing a little work on ", x)
        time.sleep(0.1)
    return x


df = daft.from_pydict({"a": [i for i in range(48)]}).with_column("b", do_work(col("a"))).agg(col("b").sum())
df.collect()
```